crypto/crypto_aesctr.c

```diff
@@ -2,18 +2,51 @@
 #include <stdint.h>
 #include <stdlib.h>
 
+#include "cpusupport.h"
 #include "crypto_aes.h"
+#include "crypto_aesctr_aesni.h"
 #include "insecure_memzero.h"
 #include "sysendian.h"
 
 #include "crypto_aesctr.h"
 
-struct crypto_aesctr {
-	const struct crypto_aes_key * key;
-	uint64_t bytectr;
-	uint8_t buf[16];
-	uint8_t pblk[16];
-};
+/**
+ * In order to optimize AES-CTR, it is desirable to separate out the handling
+ * of individual bytes of data vs. the handling of complete (16 byte) blocks.
+ * The handling of blocks in turn can be optimized further using CPU
+ * intrinsics, e.g. SSE2 on x86 CPUs; however while the byte-at-once code
+ * remains the same across platforms it should be inlined into the same (CPU
+ * feature specific) routines for performance reasons.
+ *
+ * In order to allow those generic functions to be inlined into multiple
+ * functions in separate translation units, we place them into a "shared" C
+ * file which is included in each of the platform-specific variants.
+ */
+#include "crypto_aesctr_shared.c"
+
+/**
+ * crypto_aes_use_x86_aesni(void):
+ * Return non-zero if AESNI operations are available.
+ */
+static int
+crypto_aesctr_use_x86_aesni(void)
+{
+	static int aesnigood = -1;
+
+#if defined(CPUSUPPORT_X86_AESNI)
```

Comment:
Why not make this entire function `#if defined(CPUSUPPORT_X86_AESNI)`?

Reply:
oh right, that construct originated from `crypto_aes_use_x86_aesni()` so that it would `return 0;` if it's not defined, but of course we don't need that any more.